src/outputs/manifest_output.py

```diff
@@ -0,0 +1,202 @@
+import json
+import os
+from datetime import datetime
+
+
+class ManifestError(Exception):
+    pass
+
+
+class Manifest:
+    """
+    An outgoing file transfer manifest. Accumulates file metadata before
+        writing a JSON manifest file.
+
+    Used by Apache NiFi for data integrity checks before initiating data
+        transfers.
+
+    Attributes
+    ==========
+    outgoing_directory
+        location to write files to
+    pipeline_run_datetime
+        datetime of the current pipeline run, used to version outputs
+    dry_run
+        when True, cleans up output files after a successful run
+    """
+
+    def __init__(
+        self,
+        outgoing_directory: str,
+        export_directory: str,
+        pipeline_run_datetime: datetime,
+        delete_file_func: callable,
+        md5sum_func: callable,
+        stat_size_func: callable,
+        isdir_func: callable,
+        isfile_func: callable,
+        read_header_func: callable,
+        string_to_file_func: callable,
+        dry_run: bool = False,
+    ):
+        self.outgoing_directory = outgoing_directory
+        self.export_directory = export_directory
+        if not isdir_func(outgoing_directory):
+            raise ManifestError(
+                f"Outgoing directory does not exist: {self.outgoing_directory}"
+            )
+        if not isdir_func(export_directory):
+            raise ManifestError(
+                f"Outgoing directory does not exist: {self.outgoing_directory}"
+            )
+
+        if not isinstance(pipeline_run_datetime, datetime):
+            raise ManifestError("Pipeline run datetime must be a datetime object.")
+
+        self.manifest_datetime = pipeline_run_datetime.strftime("%Y%m%d_%H%M%S")
```

Comment:
The output file name is hard to read (20230928_171230_manifest.json). Can you change the format of the datetime slightly to make it more readable especially for users?

Reply:
This is a difficult one. I agree that's not particularly user friendly but I need it to be something that is unique. As the manifest can contain metadata for many csv files, I can't use a single file name as the basis for the manifest file name.

Reply:
how about YYYY-MM-DD_etc ?

Reply:
I have updated the name. I kept in there the time because it makes it more unique. Having the time in there with a ":" in was causing the bug in hdfs though! I think it's sorted now. 

Also, I don't think the manifest file needs to be very user-readable, as it's for nifi to pick up.